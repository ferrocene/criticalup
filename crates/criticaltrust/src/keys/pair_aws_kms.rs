// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

use super::newtypes::{PayloadBytes, SignatureBytes};
use super::PublicKey;
use crate::keys::newtypes::PublicKeyBytes;
use crate::keys::{KeyAlgorithm, KeyPair, KeyRole};
use crate::sha256::hash_sha256;
use crate::Error;
use aws_sdk_kms::primitives::Blob;
use aws_sdk_kms::types::{KeySpec, MessageType, SigningAlgorithmSpec};
use aws_sdk_kms::Client;
use tokio::runtime::Handle;

/// Pair of public and private keys stored in [AWS KMS](https://aws.amazon.com/kms/).
///
/// The private key is exclusively stored inside of AWS KMS, and this struct makes network calls to
/// kMS for every signature request. The public key is downloaded locally when the struct is
/// instantiated, and signatures are verified without making network calls.
pub struct AwsKmsKeyPair {
    handle: Handle,
    kms: Client,
    key_id: String,
    public_key: PublicKey,
}

impl AwsKmsKeyPair {
    /// Load an AWS KMS asymmetric key. The key must exist, and must use one of the algorithms
    /// supported by criticaltrust.
    pub fn new(
        key_id: &str,
        tokio_handle: Handle,
        kms_client: Client,
        role: KeyRole,
    ) -> Result<Self, Error> {
        let public_key_response =
            tokio_handle.block_on(kms_client.get_public_key().key_id(key_id).send())?;

        let public_key = match public_key_response.key_spec() {
            Some(KeySpec::EccNistP256)
                if public_key_response
                    .signing_algorithms()
                    .contains(&SigningAlgorithmSpec::EcdsaSha256) =>
            {
                PublicKey {
                    role,
                    algorithm: KeyAlgorithm::EcdsaP256Sha256Asn1SpkiDer,
                    expiry: None,
                    public: PublicKeyBytes::owned(
                        public_key_response
                            .public_key()
                            .unwrap()
                            .clone()
                            .into_inner(),
                    ),
                }
            }
            _ => return Err(Error::UnsupportedKey),
        };

        Ok(Self {
            handle: tokio_handle,
            kms: kms_client,
            key_id: key_id.into(),
            public_key,
        })
    }
}

impl KeyPair for AwsKmsKeyPair {
    fn public(&self) -> &PublicKey {
        &self.public_key
    }

    fn sign(&self, data: &PayloadBytes<'_>) -> Result<SignatureBytes<'static>, Error> {
        let (digest, algorithm) = match self.public_key.algorithm {
            KeyAlgorithm::EcdsaP256Sha256Asn1SpkiDer => (
                hash_sha256(data.as_bytes()),
                SigningAlgorithmSpec::EcdsaSha256,
            ),
            KeyAlgorithm::Unknown => return Err(Error::UnsupportedKey),
        };

        let response = self.handle.block_on(
            self.kms
                .sign()
                .key_id(&self.key_id)
                .message(Blob::new(digest))
                .message_type(MessageType::Digest)
                .signing_algorithm(algorithm)
                .send(),
        )?;

        Ok(SignatureBytes::owned(
            response.signature().unwrap().clone().into_inner(),
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use aws_sdk_kms::config::Credentials;
    use aws_sdk_kms::types::KeyUsageType;
    use rand_core::{OsRng, RngCore};
    use std::process::Command;
    use std::sync::Once;
    use tokio::runtime::Runtime;

    // We want to have tests for all of criticaltrust, which makes testing the integration with AWS
    // KMS quite tricky. To make it work, the tests for this module spawn a Docker container for
    // "localstack", a local replica of AWS services meant for testing.

    #[test]
    fn test_roundtrip() {
        let localstack = Localstack::init();
        let key = localstack.create_key(KeySpec::EccNistP256);

        let keypair = AwsKmsKeyPair::new(
            &key,
            localstack.runtime.handle().clone(),
            localstack.client.clone(),
            KeyRole::Root,
        )
        .expect("failed to create key pair");

        let payload = PayloadBytes::borrowed(b"Hello world");
        let signature = keypair.sign(&payload).expect("failed to sign");
        keypair
            .public()
            .verify(KeyRole::Root, &payload, &signature, None)
            .expect("failed to verify");
    }

    #[test]
    fn test_key_pair_with_unsupported_algorithm() {
        let localstack = Localstack::init();
        let key = localstack.create_key(KeySpec::Rsa2048);

        let keypair = AwsKmsKeyPair::new(
            &key,
            localstack.runtime.handle().clone(),
            localstack.client.clone(),
            KeyRole::Root,
        );
        assert!(matches!(keypair, Err(Error::UnsupportedKey)));
    }

    struct Localstack {
        runtime: Runtime,
        client: Client,
        container_name: String,
    }

    impl Localstack {
        fn init() -> Self {
            let image = pull_localstack_docker_image();
            let container_name = format!("criticaltrust-localstack-{}", OsRng.next_u64());

            run(Command::new("docker")
                .arg("create")
                .args(["--name", &container_name])
                .args(["-p", "4566"])
                .arg(image));
            run(Command::new("docker").args(["start", &container_name]));

            // localstack is bound to a random port to prevent conflicts between concurrent tests.
            // We thus need to fetch the actual port number Docker assigned.
            let ports = run(Command::new("docker").args(["port", &container_name, "4566/tcp"]));
            let port = std::str::from_utf8(&ports)
                .expect("non-utf-8 output of docker port")
                .split('\n')
                .next()
                .expect("empty output of docker port")
                .rsplit_once(':')
                .expect("invalid output of docker port")
                .1;

            let runtime = Runtime::new().expect("failed to create tokio runtime");
            let aws_config = runtime.block_on(
                aws_config::from_env()
                    // localstack doesn't validate IAM credentials, so we can configure a dummy
                    // secret key and region.
                    .credentials_provider(Credentials::new(
                        "aws_access_key_id",
                        "aws_secret_access_key",
                        None,
                        None,
                        "hardcoded",
                    ))
                    .region("us-east-1")
                    .load(),
            );

            let kms_config = aws_sdk_kms::config::Builder::from(&aws_config)
                .endpoint_url(format!("http://localhost:{port}"))
                .build();
            let client = aws_sdk_kms::Client::from_conf(kms_config);

            Self {
                runtime,
                client,
                container_name,
            }
        }

        fn create_key(&self, spec: KeySpec) -> String {
            self.runtime
                .block_on(
                    self.client
                        .create_key()
                        .key_usage(KeyUsageType::SignVerify)
                        .key_spec(spec)
                        .send(),
                )
                .expect("failed to create kms key")
                .key_metadata()
                .unwrap()
                .key_id()
                .into()
        }
    }

    impl Drop for Localstack {
        fn drop(&mut self) {
            run(Command::new("docker").args(["stop", &self.container_name, "-t", "10"]));
            run(Command::new("docker").args(["rm", &self.container_name]));
        }
    }

    fn pull_localstack_docker_image() -> &'static str {
        const IMAGE: &str = "localstack/localstack:2.2.0";
        static ONCE: Once = Once::new();

        ONCE.call_once(|| {
            run(Command::new("docker").args(["pull", IMAGE]));
        });

        IMAGE
    }

    fn run(command: &mut Command) -> Vec<u8> {
        let repr = format!("{command:?}");
        eprintln!("running {repr}");
        match command.output() {
            Ok(output) if output.status.success() => {
                eprintln!("finished running {repr}");
                output.stdout
            }
            Ok(output) => {
                eprintln!("failed to run command: exited with {}", output.status);
                eprintln!("command: {repr}");
                eprintln!(
                    "\nstdout:\n=====\n{}\n=====",
                    String::from_utf8_lossy(&output.stdout)
                );
                eprintln!(
                    "\nstderr:\n=====\n{}\n=====",
                    String::from_utf8_lossy(&output.stderr)
                );
                panic!();
            }
            Err(err) => panic!("command failed to start ({err}: {repr})"),
        }
    }
}
