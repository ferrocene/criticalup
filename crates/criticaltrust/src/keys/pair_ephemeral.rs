// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

use crate::keys::newtypes::{PayloadBytes, PrivateKeyBytes, SignatureBytes};
use crate::keys::{KeyAlgorithm, KeyId, KeyPair, KeyRole, PublicKey};
use crate::signatures::PublicKeysRepository;
use crate::Error;
use time::OffsetDateTime;

/// Pair of public and private keys generated at runtime and kept in memory.
///
/// There is intentionally no way to persist the private key of ephemeral key pairs, as that's
/// considerably less secure than storing the key in a Hardware Security Module. Ephemeral key
/// pairs are primarily meant to be used during automated testing.
pub struct EphemeralKeyPair {
    public: PublicKey,
    private: PrivateKeyBytes<'static>,
}

impl PublicKeysRepository for EphemeralKeyPair {
    fn get<'a>(&'a self, _id: &KeyId) -> Option<&'a PublicKey> {
        Some(&self.public)
    }
}

impl EphemeralKeyPair {
    /// Generate a new key pair using the given algorithm and key role.
    pub fn generate(
        algorithm: KeyAlgorithm,
        role: KeyRole,
        expiry: Option<OffsetDateTime>,
    ) -> Result<Self, Error> {
        let private = algorithm.methods().generate_private_key()?;

        Ok(EphemeralKeyPair {
            public: PublicKey {
                role,
                algorithm,
                expiry,
                public: algorithm
                    .methods()
                    .derive_public_key_from_private_key(&private)?,
            },
            private,
        })
    }
}

impl KeyPair for EphemeralKeyPair {
    fn public(&self) -> &PublicKey {
        &self.public
    }

    fn sign(&self, data: &PayloadBytes<'_>) -> Result<SignatureBytes<'static>, Error> {
        self.public.algorithm.methods().sign(&self.private, data)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const ALGORITHM: KeyAlgorithm = KeyAlgorithm::EcdsaP256Sha256Asn1SpkiDer;

    #[test]
    fn test_unique_keys_are_generated() {
        assert_ne!(
            EphemeralKeyPair::generate(ALGORITHM, KeyRole::Root, None)
                .unwrap()
                .private,
            EphemeralKeyPair::generate(ALGORITHM, KeyRole::Root, None)
                .unwrap()
                .private
        );
    }

    #[test]
    fn test_key_with_unknown_algorithm_is_not_generated() {
        assert!(matches!(
            EphemeralKeyPair::generate(KeyAlgorithm::Unknown, KeyRole::Root, None),
            Err(Error::UnsupportedKey),
        ));
    }

    #[test]
    fn test_public_key_is_expected() {
        let key = EphemeralKeyPair::generate(ALGORITHM, KeyRole::Root, None).unwrap();

        let public = key.public();
        assert_eq!(public.role, KeyRole::Root);
        assert_eq!(public.algorithm, ALGORITHM);
        assert_eq!(
            public.public,
            ALGORITHM
                .methods()
                .derive_public_key_from_private_key(&key.private)
                .unwrap()
        );
    }

    #[test]
    fn test_signatures_are_valid() {
        let key = EphemeralKeyPair::generate(ALGORITHM, KeyRole::Root, None).unwrap();
        let data = PayloadBytes::borrowed(b"Hello world");

        // We can't verify the exact signature is what we expect, as each signature includes random
        // data in it. Instead, we ensure it's correct.
        let signature = key.sign(&data).unwrap();

        assert!(ALGORITHM
            .methods()
            .verify(&key.public.public, &data, &signature)
            .is_ok());
    }

    #[test]
    fn test_sign_with_unknown_algorithm_fails() {
        let mut key = EphemeralKeyPair::generate(ALGORITHM, KeyRole::Root, None).unwrap();
        key.public.algorithm = KeyAlgorithm::Unknown;

        assert!(matches!(
            key.sign(&PayloadBytes::borrowed(b"Hello world")),
            Err(Error::UnsupportedKey)
        ));
    }
}
