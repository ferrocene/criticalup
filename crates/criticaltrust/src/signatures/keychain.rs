// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

use crate::keys::{KeyId, KeyRole, PublicKey};
use crate::manifests::KeysManifest;
use crate::revocation_info::RevocationInfo;
use crate::signatures::{PublicKeysRepository, SignedPayload};
use crate::Error;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Collection of all trusted public keys.
#[derive(Serialize, Deserialize)]
pub struct Keychain {
    keys: HashMap<KeyId, PublicKey>,
    revocation_info: Option<RevocationInfo>,
}

impl Keychain {
    /// Create a new keychain, using the provided public key as the root of trust.
    ///
    /// The root of trust has to have the `root` key role, and all future keys added to the
    /// keychain will have to be signed by either the root of trust or another key signed by the
    /// root of trust.
    pub fn new(trust_root: &PublicKey) -> Result<Self, Error> {
        let mut keychain = Self {
            keys: HashMap::new(),
            revocation_info: None,
        };

        if trust_root.role != KeyRole::Root {
            return Err(Error::WrongKeyRoleForTrustRoot(trust_root.role));
        }
        keychain.load_inner(trust_root)?;

        Ok(keychain)
    }

    pub fn keys(&self) -> &HashMap<KeyId, PublicKey> {
        &self.keys
    }

    pub fn revocation_info(&self) -> Option<&RevocationInfo> {
        self.revocation_info.as_ref()
    }

    /// Update the [`Keychain`] provided the [`KeysManifest`]:
    /// 1. Verify and load all the verified keys.
    /// 2. Verify and replace the Revocation information from the revoked content.
    pub fn load_all(&mut self, keys_manifest: &KeysManifest) -> Result<(), Error> {
        if self.revocation_info.is_some() {
            return Err(Error::RevocationInfoOverwriting);
        }

        // Load all keys from KeysManifest.
        for key in &keys_manifest.keys {
            // Invalid keys are silently ignored, as they might be signed by a different root key
            // used by a different release of criticalup, or they might be using an algorithm not
            // supported by the current version of criticaltrust.
            let _ = self.load(key)?;
        }

        // Special case: verify and load only RevocationInfo.
        let revocation_info = keys_manifest.revoked_signatures.get_verified(self)?;
        self.revocation_info = Some(revocation_info.clone());

        Ok(())
    }

    /// Add a new signed key to the keychain.
    ///
    /// The key has to be signed by either the root of trust or another key with the root role
    /// already part of the keychain.
    pub fn load(&mut self, key: &SignedPayload<PublicKey>) -> Result<KeyId, Error> {
        let key = key.get_verified(self)?;
        self.load_inner(&key)
    }

    fn load_inner(&mut self, key: &PublicKey) -> Result<KeyId, Error> {
        if !key.is_supported() {
            return Err(Error::UnsupportedKey);
        }
        let id = key.calculate_id();
        self.keys.insert(id.clone(), key.clone());
        Ok(id)
    }
}

impl PublicKeysRepository for Keychain {
    fn get(&self, id: &KeyId) -> Option<&PublicKey> {
        self.keys.get(id)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::keys::{EphemeralKeyPair, KeyAlgorithm, KeyPair};
    use crate::manifests::ManifestVersion;
    use crate::signatures::{Signable, SignedPayload};
    use time::macros::datetime;
    use time::{Duration, OffsetDateTime};

    // Make sure there is enough number of days for expiration so tests don't need constant updates.
    const EXPIRATION_EXTENSION_IN_DAYS: Duration = Duration::days(180);

    #[test]
    fn test_new_with_root_key_as_trust_root() {
        let root = generate_key(KeyRole::Root);

        let keychain = Keychain::new(root.public()).unwrap();
        assert_eq!(
            Some(root.public()),
            keychain.get(&root.public().calculate_id())
        );
    }

    #[test]
    fn test_new_with_non_root_key_as_trust_root() {
        let non_root = generate_key(KeyRole::Packages);

        assert!(matches!(
            Keychain::new(non_root.public()),
            Err(Error::WrongKeyRoleForTrustRoot(KeyRole::Packages))
        ));
    }

    #[tokio::test]
    async fn test_add_key_trusted_by_root() {
        let root = generate_key(KeyRole::Root);
        let mut keychain = Keychain::new(root.public()).unwrap();

        let (key, public) = generate_trusted_key(KeyRole::Packages, &root).await;
        keychain.load(&public).unwrap();

        assert_eq!(
            Some(key.public()),
            keychain.get(&key.public().calculate_id())
        );
    }

    #[tokio::test]
    async fn test_add_key_trusted_by_root_key_trusted_by_root() {
        let root = generate_key(KeyRole::Root);
        let mut keychain = Keychain::new(root.public()).unwrap();

        let (key1, public1) = generate_trusted_key(KeyRole::Root, &root).await;
        keychain.load(&public1).unwrap();

        let (key2, public2) = generate_trusted_key(KeyRole::Packages, &key1).await;
        keychain.load(&public2).unwrap();

        assert_eq!(
            Some(key2.public()),
            keychain.get(&key2.public().calculate_id())
        );
    }

    #[tokio::test]
    async fn test_add_key_trusted_by_non_root_key_trusted_by_root() {
        let root = generate_key(KeyRole::Root);
        let mut keychain = Keychain::new(root.public()).unwrap();

        let (key1, public1) = generate_trusted_key(KeyRole::Packages, &root).await;
        keychain.load(&public1).unwrap();

        let (_, public2) = generate_trusted_key(KeyRole::Packages, &key1).await;
        assert!(matches!(
            keychain.load(&public2),
            Err(Error::VerificationFailed)
        ));
    }

    #[tokio::test]
    async fn test_add_key_trusted_by_nothing_else() {
        let mut keychain = Keychain::new(generate_key(KeyRole::Root).public()).unwrap();

        let another_root = generate_key(KeyRole::Root);
        let (_, public) = generate_trusted_key(KeyRole::Packages, &another_root).await;
        assert!(matches!(
            keychain.load(&public),
            Err(Error::VerificationFailed)
        ));
    }

    #[tokio::test]
    async fn test_add_key_with_unsupported_algorithm() {
        let root = generate_key(KeyRole::Root);
        let mut keychain = Keychain::new(root.public()).unwrap();

        let mut other: SignedPayload<PublicKey> = SignedPayload::new(
            &serde_json::from_str(
                r#"{"algorithm": "foo", "role": "root", "expiry": null, "public": "aGk="}"#,
            )
            .unwrap(),
        )
        .unwrap();
        other.add_signature(&root).await.unwrap();

        assert!(matches!(keychain.load(&other), Err(Error::UnsupportedKey)));
    }

    impl Signable for String {
        const SIGNED_BY_ROLE: KeyRole = KeyRole::Packages;
    }

    impl Signable for Vec<u8> {
        const SIGNED_BY_ROLE: KeyRole = KeyRole::Revocation;
    }

    // Test `load_all` method with RevocationInfo being an empty list.
    #[tokio::test]
    async fn test_load_all_revoked_content_empty() {
        let root = generate_key(KeyRole::Root);
        let (revocation_keypair, signed_public_revocation_key) =
            generate_trusted_key(KeyRole::Revocation, &root).await;

        let revoked_content = RevocationInfo::new(vec![], datetime!(2400-10-10 00:00 UTC));
        let mut signed_revoked_content = SignedPayload::new(&revoked_content).unwrap();
        signed_revoked_content
            .add_signature(&revocation_keypair)
            .await
            .unwrap();

        let mut keychain = Keychain::new(root.public()).unwrap();

        let keys_manifest = KeysManifest {
            version: ManifestVersion,
            keys: vec![signed_public_revocation_key],
            revoked_signatures: signed_revoked_content,
        };

        keychain.load_all(&keys_manifest).unwrap();
        assert_eq!(
            keychain.revocation_info.unwrap().expires_at,
            datetime!(2400-10-10 00:00 UTC)
        )
    }

    // Test `load_all` method with RevocationInfo but with one item in the list. The call
    // to `load_all` should not fail in verifying the revocation key.
    #[tokio::test]
    async fn test_load_all_revoked_content_one_item() {
        let root = generate_key(KeyRole::Root);
        let (revocation_keypair, signed_public_revocation_key) =
            generate_trusted_key(KeyRole::Revocation, &root).await;
        let revoked_content = RevocationInfo::new(
            vec![vec![1, 2, 3]],
            OffsetDateTime::now_utc() + EXPIRATION_EXTENSION_IN_DAYS,
        );
        let mut signed_revoked_content = SignedPayload::new(&revoked_content).unwrap();
        signed_revoked_content
            .add_signature(&revocation_keypair)
            .await
            .unwrap();

        let mut keychain = Keychain::new(root.public()).unwrap();

        let keys_manifest = KeysManifest {
            version: ManifestVersion,
            keys: vec![signed_public_revocation_key],
            revoked_signatures: signed_revoked_content,
        };

        keychain.load_all(&keys_manifest).unwrap();
        let binding = keychain.revocation_info.unwrap();
        let actual = binding.revoked_content_sha256.first().unwrap();
        let expected: &Vec<u8> = &vec![1, 2, 3];
        assert_eq!(actual, expected);
    }

    #[tokio::test]
    async fn test_error_on_load_all_when_revocation_info_is_some() {
        let root = generate_key(KeyRole::Root);
        let (revocation_keypair, signed_public_revocation_key) =
            generate_trusted_key(KeyRole::Revocation, &root).await;
        let revoked_content = RevocationInfo::new(
            vec![vec![1, 2, 3]],
            OffsetDateTime::now_utc() + EXPIRATION_EXTENSION_IN_DAYS,
        );
        let mut signed_revoked_content = SignedPayload::new(&revoked_content).unwrap();
        signed_revoked_content
            .add_signature(&revocation_keypair)
            .await
            .unwrap();
        let mut keychain = Keychain::new(root.public()).unwrap();
        let keys_manifest = KeysManifest {
            version: ManifestVersion,
            keys: vec![signed_public_revocation_key],
            revoked_signatures: signed_revoked_content,
        };
        keychain.load_all(&keys_manifest).unwrap();
        assert!(keychain.revocation_info.is_some());
        assert!(matches!(
            keychain.load_all(&keys_manifest),
            Err(Error::RevocationInfoOverwriting)
        ));
    }

    // Utilities

    fn generate_key(role: KeyRole) -> EphemeralKeyPair {
        EphemeralKeyPair::generate(KeyAlgorithm::EcdsaP256Sha256Asn1SpkiDer, role, None).unwrap()
    }

    async fn generate_trusted_key<K: KeyPair>(
        role: KeyRole,
        trusted_by: &K,
    ) -> (EphemeralKeyPair, SignedPayload<PublicKey>) {
        let key = generate_key(role);
        let mut payload = SignedPayload::new(key.public()).unwrap();
        payload.add_signature(trusted_by).await.unwrap();
        (key, payload)
    }
}
