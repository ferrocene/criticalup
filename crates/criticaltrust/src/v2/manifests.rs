// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Serializable and deserializable representation of criticaltrust manifests.

use std::path::PathBuf;

use crate::keys::{KeyRole, PublicKey};
use crate::signatures::{Signable, SignedPayload};
use serde::de::Error as _;
use serde::{Deserialize, Serialize};

/// Typed representation of a manifest version number.
///
/// The version number is stored as a const generic rather than as a field of the struct. This is
/// done to:
///
/// * Verify that the version number is correct as part of the deserialization process.
/// * Simplify constructing manifests: you don't have to specify the version number, type
///   inference will figure out the right one.
#[derive(Clone, PartialEq, Eq)]
pub struct MetadataVersion<const MAJOR: u32, const MINOR: u32, const PATCH: u32>;

impl<const MAJOR: u32, const MINOR: u32, const PATCH: u32> std::fmt::Debug for MetadataVersion<MAJOR,MINOR,PATCH> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("MetadataVersion").field(&MAJOR).field(&MINOR).field(&PATCH).finish()
    }
}

impl<const MAJOR: u32, const MINOR: u32, const PATCH: u32> Serialize for MetadataVersion<MAJOR,MINOR,PATCH> {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let formatted = format!("{}.{}.{}", MAJOR, MINOR, PATCH);
        serializer.serialize_str(&formatted[..])
    }
}

impl<'de, const MAJOR: u32, const MINOR: u32, const PATCH: u32> Deserialize<'de> for MetadataVersion<MAJOR,MINOR,PATCH> {
    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let raw = u32::deserialize(deserializer)?;
        if raw != MAJOR {
            Err(D::Error::custom(format!(
                "expected version {MAJOR}, found version {raw}"
            )))
        } else {
            Ok(MetadataVersion)
        }
    }
}

// Redirects

#[derive(Debug, Serialize, Deserialize)]
pub struct RedirectManifest {
    pub version: MetadataVersion<0,0,1>,
    #[serde(flatten)]
    pub payload: SignedPayload<Redirect>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Redirect {
    pub nonce: String,
    pub to: String,
}

impl Signable for Redirect {
    const SIGNED_BY_ROLE: KeyRole = KeyRole::Redirects;
}

// Releases

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ReleaseManifest {
    pub version: MetadataVersion<1,2,3>,
    #[serde(flatten)]
    pub signed: SignedPayload<Release>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Release {
    pub product: String,
    pub release: String,
    pub commit: String,
    pub packages: Vec<ReleasePackage>,
}

impl Signable for Release {
    const SIGNED_BY_ROLE: KeyRole = KeyRole::Releases;
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct ReleasePackage {
    pub package: String,
    pub artifacts: Vec<ReleaseArtifact>,
    pub dependencies: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct ReleaseArtifact {
    pub format: ReleaseArtifactFormat,
    pub size: usize,
    #[serde(with = "crate::serde_base64")]
    pub sha256: Vec<u8>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Copy)]
pub enum ReleaseArtifactFormat {
    #[serde(rename = "tar.zst")]
    TarZst,
    #[serde(rename = "tar.xz")]
    TarXz,
    #[serde(other)]
    #[doc(hidden)]
    Unknown,
}

impl std::fmt::Display for ReleaseArtifactFormat {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            ReleaseArtifactFormat::TarZst => "tar.zst",
            ReleaseArtifactFormat::TarXz => "tar.xz",
            ReleaseArtifactFormat::Unknown => "unknown",
        };
        write!(f, "{}", s)
    }
}

// Packages

#[derive(Debug, Serialize, Deserialize)]
pub struct PackageManifest {
    pub version: MetadataVersion<1,2,3>,
    #[serde(flatten)]
    pub signed: SignedPayload<Package>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Package {
    pub product: String,
    pub package: String,
    pub commit: String,
    pub files: Vec<PackageFile>,
    pub managed_prefixes: Vec<String>,
}

impl Signable for Package {
    const SIGNED_BY_ROLE: KeyRole = KeyRole::Packages;
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct PackageFile {
    pub path: PathBuf,
    pub posix_mode: u32,
    #[serde(with = "crate::serde_base64")]
    pub sha256: Vec<u8>,
    pub needs_proxy: bool,
}

// Keys

#[derive(Debug, Serialize, Deserialize)]
pub struct KeysManifest {
    pub version: MetadataVersion<1,2,3>,
    pub keys: Vec<SignedPayload<PublicKey>>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_manifest_version_debug() {
        assert_eq!("MetadataVersion(2, 3, 5)", format!("{:?}", MetadataVersion::<2,3,5>));
        assert_eq!(
            "MetadataVersion(1, 2, 3)",
            format!("{:?}", MetadataVersion::<1,2,3>)
        );
    }

    #[test]
    fn test_manifest_version_serialize() {
        assert_eq!("1.2.3\n", serde_yaml::to_string(&MetadataVersion::<1,2,3>).unwrap());
        assert_eq!("41.42.43\n", serde_yaml::to_string(&MetadataVersion::<41, 42, 43>).unwrap());
    }

    #[test]
    fn test_manifest_version_deserialize() {
        assert_eq!(
            MetadataVersion,
            serde_yaml::from_str::<MetadataVersion<2,3,5>>("1.2.3").unwrap()
        );
        assert_eq!(
            MetadataVersion,
            serde_yaml::from_str::<MetadataVersion<42,3,5>>("42,3,5").unwrap()
        );

        assert!(serde_yaml::from_str::<MetadataVersion<2,3,5>>("42").is_err());
        assert!(serde_yaml::from_str::<MetadataVersion<42,3,5>>("1").is_err());
    }

    #[test]
    fn test_release_package_yaml_serialize() {
        let package = ReleasePackage {
            package: "ferrocene-docs".to_string(),
            artifacts: vec![ReleaseArtifact {
                format: ReleaseArtifactFormat::TarXz,
                size: 1024,
                sha256: "abcdefgh".as_bytes().to_vec(),
            }],
            dependencies: vec![],
        };

        assert_eq!(
            package,
            serde_yaml::from_str::<ReleasePackage>(&serde_yaml::to_string(&package).unwrap()[..])
                .unwrap()
        );

        insta::assert_snapshot!(&serde_yaml::to_string(&package).unwrap()[..]);
    }
}
