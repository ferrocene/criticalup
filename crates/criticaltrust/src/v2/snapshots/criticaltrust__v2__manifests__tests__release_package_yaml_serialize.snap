---
source: crates/criticaltrust/src/v2/manifests.rs
expression: "&serde_yaml::to_string(&package).unwrap()[..]"
---
package: ferrocene-docs
artifacts:
- format: tar.xz
  size: 1024
  sha256: YWJjZGVmZ2g=
dependencies: []
